File: Geometry/Lines_and_Queries_II.py
import sys,math,cmath,random,os
# from heapq import heappush,heappop
# from bisect import bisect_right,bisect_left
# from collections import Counter,deque,defaultdict
# from itertools import permutations,combinations
from io import BytesIO, IOBase
# from decimal import Decimal,getcontext

BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)

# functions #
# MOD = 998244353
# MOD = 10**9 + 7
# RANDOM = random.randrange(1,2**62)
# def gcd(a,b):
#     while b:
#         a,b = b,a%b
#     return a
# def lcm(a,b):
#     return a//gcd(a,b)*b
# def w(x):
#     return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
# MI = lambda x : x(map(int, sys.stdin.readline().split()))
# SI = lambda : sys.stdin.readline().strip()
# SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
# LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
# LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
# MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##

#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, bootstrap: bootstrap, Trie: tries, Segment Tree(lp): SegmentOther, Treap: treap
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, bwt: bwt, DynamicConnectivity: odc
#Graph1(axtree,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, matrix: nummat, Suffix/KMPAutomaton: sautomaton, linalg: linalg, SquareRtDecomp: sqrt
#Grapth7(bridges): graph_dmgt, FWHT(^,|,&): fwht, Graph8(centr_decom): graph_decom, DpOptimize(knth,dnc): dpopt
#2D-SegmentTree: seg2d, Rollback/Par DSU: rbdsu, MaxFlow(Dnc,HLPP): graphflow, MaxMatching(Kuhn,Hopcroft): graphmatch
#Segment Tree(Node): SegmentNode, mcmf: mcmf
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
#if os.environ.get('LOCAL'):sys.stdin = open(r'input.txt', 'r');sys.stdout = sys.stderr = open(r'output.txt','w')
#if os.environ.get('LOCAL'):import hashlib;print('Hash Value :',hashlib.md5(open(__file__, 'rb').read()).hexdigest());

class LiChao:
    __slots__ = ("lo","hi","mid","m","b","left","right")
    def __init__(self,lo,hi):
        self.lo = lo
        self.hi = hi
        self.mid = (lo+hi)>>1
        self.m = None
        self.b = None
        self.left = None
        self.right = None
    def add_line(self,m,b):
        node = self
        while True:
            lo,hi,mid = node.lo,node.hi,node.mid
            if node.m is None:
                node.m,node.b = m,b
                return
            cur_m,cur_b = node.m,node.b
            if m*mid+b>cur_m*mid+cur_b:
                node.m,m = m,cur_m
                node.b,b = b,cur_b
                cur_m,cur_b = node.m,node.b
            if lo==hi:
                return
            if m*lo+b>cur_m*lo+cur_b:
                if node.left is None:
                    node.left = LiChao(lo,mid)
                node = node.left
            elif m*hi+b>cur_m*hi +cur_b:
                if node.right is None:
                    node.right = LiChao(mid+1,hi)
                node = node.right
            else:
                return
    def query(self,x):
        node = self
        res = -float('inf')
        while node:
            if node.m is not None:
                v = node.m*x+node.b
                if v>res:
                    res = v
            if node.lo==node.hi:
                break
            if x<=node.mid:
                node = node.left
            else:
                node = node.right
        return res


        # self.size = 1 << (self.n - 1).bit_length()
size = 1<<17
tree = [None for i in range (2*size)]
def range_update(left, right, value):
    # Range Update in [L,R] if flag, then add
    l = left + size
    r = right + size
    while l <= r:
        # print('Y',l,r)
        if l & 1: 
            if not tree[l]:
                tree[l] = LiChao(0,1<<17)
            tree[l].add_line(*value)
            l += 1
        if not r & 1: 
            if not tree[r]:
                tree[r] = LiChao(0,1<<17)
            tree[r].add_line(*value)
            r -= 1
        l >>= 1; r >>= 1
def range_query(pos):
    # Range Query in [L,R]
    x = pos+1
    pos += size
    ans = tree[pos].query(x) if tree[pos] else -float('inf')
    while pos>1:
        pos >>= 1
        ans = max(ans,tree[pos].query(x) if tree[pos] else -float('inf'))
    return ans

def solve():
    for _ in range(II()):
        L = LII()
        if L[0]==1:
            _,a,b,l,r = L
            range_update(l-1,r-1,(a,b))
        else:
            ans = (range_query(L[1]-1))
            print(ans if ans!=-float('inf') else "NO")
    #L1 = LII()
    #st = SI()
solve()

File: Geometry/Lines_and_Queries_II.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
// #include <ext/pb_ds/assoc_container.hpp>
// #include <ext/pb_ds/tree_policy.hpp>
// using namespace __gnu_pbds;
using namespace std;

#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define int long long
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vi>;
using viii = vector<vii>;
using ll = long long;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")
mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());

#ifdef LOCAL
#define dbg(x...) cout<<"["<<#x<<"] = "; print(x);
#else
#define dbg(x...)
#endif

// template <typename T>
// using ordered_multiset = tree<
//     std::pair<T, int>,
//     null_type,
//     std::less<std::pair<T, int>>,
//     rb_tree_tag,
//     tree_order_statistics_node_update
// >;
template<class T> bool chmin(T& a,const T& b){ if(b<a){ a=b; return 1;} return 0; }
template<class T> bool chmax(T& a,const T& b){ if(a<b){ a=b; return 1;} return 0;}
template<class T> using minpq = priority_queue<T,vector<T>,greater<T>>;
template<class T> using maxpq = priority_queue<T>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<class T> void _print(const T &x){ cout << x; }
template<class T,class U> void _print(const pair<T,U> &p){cout<<"(";_print(p.first);cout<<",";_print(p.second);cout<<")";}
template<class T>
void _print(const vector<T> &v){
    cout<<"[";bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"]";
}
template<class T>
void _print(const set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T>
void _print(const unordered_set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const unordered_map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
void print(){ cout<<"\n"; }
template<class T,class... Args>
void print(const T& a,const Args&... args){
    _print(a);
    if constexpr(sizeof...(args)) cout<<" ";
    print(args...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}
int rnd(int l,int r){return uniform_int_distribution<int>(l,r)(rng);}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable, SortedList: sortl
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree, Trie/Treap: Tries
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, Centroid Decomp.: graph_decom
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, GrapthOth: graphoth, DSU: DSU, FFT:fft
// Rollback/Par DSU: rbdsu, treap: treap, graphflow(mat_match): graphflow, Persistent Seg Tree: perseg
// Segment Tree(Nodes): SegmentNode, HLD: hld, fwht: fwht
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

struct LiChao {
    int lo, hi, mid;
    int m, b;
    bool has;
    LiChao *left, *right;

    LiChao(int lo, int hi) : lo(lo), hi(hi), mid((lo + hi) >> 1),m(0), b(0), has(false),left(nullptr), right(nullptr) {}

    int f(int m, int b, int x) {
        return m * x + b;
    }

    void add_line(int nm, int nb) {
        LiChao* node = this;
        while (true) {
            if (!node->has) {
                node->m = nm;
                node->b = nb;
                node->has = true;
                return;
            }

            int lo = node->lo, hi = node->hi, mid = node->mid;
            int cm = node->m, cb = node->b;

            if (f(nm, nb, mid) > f(cm, cb, mid)) {
                swap(node->m, nm);
                swap(node->b, nb);
                cm = node->m;
                cb = node->b;
            }

            if (lo == hi) return;

            if (f(nm, nb, lo) > f(cm, cb, lo)) {
                if (!node->left)
                    node->left = new LiChao(lo, mid);
                node = node->left;
            } else if (f(nm, nb, hi) > f(cm, cb, hi)) {
                if (!node->right)
                    node->right = new LiChao(mid + 1, hi);
                node = node->right;
            } else {
                return;
            }
        }
    }

    int query(int x) {
        LiChao* node = this;
        int res = LLONG_MIN;

        while (node) {
            if (node->has)
                res = max(res, f(node->m, node->b, x));

            if (node->lo == node->hi) break;

            if (x <= node->mid)
                node = node->left;
            else
                node = node->right;
        }
        return res;
    }
};

const int SIZE = 1 << 17;
LiChao* tree[2 * SIZE];

void range_update(int left, int right, pair<int,int> val) {
    int l = left + SIZE;
    int r = right + SIZE;

    while (l <= r) {
        if (l & 1) {
            if (!tree[l])
                tree[l] = new LiChao(0, 1 << 17);
            tree[l]->add_line(val.first, val.second);
            l++;
        }
        if (!(r & 1)) {
            if (!tree[r])
                tree[r] = new LiChao(0, 1 << 17);
            tree[r]->add_line(val.first, val.second);
            r--;
        }
        l >>= 1;
        r >>= 1;
    }
}

int range_query(int pos) {
    int x = pos + 1;
    pos += SIZE;

    int ans = tree[pos] ? tree[pos]->query(x) : LLONG_MIN;

    while (pos > 1) {
        pos >>= 1;
        if (tree[pos])
            ans = max(ans, tree[pos]->query(x));
    }
    return ans;
}

void solve() {
    int q;
    cin>>q;
    while(q--){
        int ty;
        cin>>ty;
        if (ty==1) {
            int a, b;
            int l, r;
            cin>>a>>b>>l>>r;
            range_update(l-1, r-1,{a, b});
        } else {
            int x;
            cin>>x;
            int ans = range_query(x-1);
            if (ans==LLONG_MIN)cout<<"NO"<<endl;
            else cout<<ans<<endl;
        }
    }
}

int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}